Evolution/NNNN-share.md

```diff
@@ -0,0 +1,178 @@
+
+```swift
+extension AsyncSequence where Element: Sendable {
+  public func share(
+    bufferingPolicy: BufferingPolicy = .unbounded
+  ) -> AsyncShareSequence<Element, Failure>
+}
+
+public struct AsyncShareSequence<Element: Sendable, Failure: Error>: AsyncSequence, Sendable {
+  public struct Iterator: AsyncIteratorProtocol {
+    public mutating func next(isolation actor: isolated (any Actor)?) async throws(Failure) -> Element? 
+  }
+
+  public func makeAsyncIterator() -> Iterator
+}
+
+@available(*, unavailable)
+extension AsyncShareSequence.Iterator: Sendable { }
+```
+
+The buffer internally to the share algorithm will only extend back to the furthest element available but there will only be a singular buffer shared across all iterators. This ensures that with the application of the buffering policy the storage size is as minimal as possible while still allowing all iterations to avoid dropping values and keeping the memory usage in check.
+
+## Runtime Behavior
+
+The construction of the `AsyncShareSequence` will initially construct a shared iteration reference. This means that all instances of the structure of the `AsyncShareSequence` will reference to the same iteration.
+
+Upon creation of the `Iterator` via `makeAsyncIterator` a new "side" will be constructed to identify the specific iterator interacting with the shared iteration. Then when next is invoked is where the first actual action takes place. 
+
+The next method will first checkout from a critical region the underlying AsyncIterator from the base. If that is successful (i.e. no other iteration sides have already checked it out) then it will invoke the next method of that iterator (forwarding in the actor isolation). If an element is produced then it enqueues the element to the shared buffer, checks in the iterator, adjusts the index in the buffer, and finds all pending continuations all in a shared critical region by a mutex. Then those continuations will be resumed with the given element.
+
+If no element is returned by the base iterator (signifying the terminal state);then the process is similar except it will instead mark the sequence as finished and resume with nil to any active continuations. Similarly with failures that will set the state as terminal but also store the error for further iteration points that need eventual termination.
+
+Then all sides are "drained" such that continuations are placed into the shared state and resumed when an element is available for that position.
+
+Practically this all means that a given iteration may be "behind" another and can eventually catch up (provided it is within the buffer limit).
```

Comment:
I think this is the reason why we need to provide the alternative variant as well that exerts back-pressure by the slowest consumer i.e. next on the base is only called once all current active iterators of share have consumed the last element. While that is most likely slower it guarantees the user that every iterator receives every element. On the other hand, with the currently proposed only buffer based solution there is no good way to set the size of the buffer that guarantees that no elements will ever be dropped.